tests/test_mutations.py

```diff
@@ -472,40 +704,226 @@ def test_keep_mutation_parent_zero_rate(self):
+                                 'new': False}
+                mutation_list.append(mutation_item)
+                j += 1
+
+            mutation_list.reverse()
+            site_item = {'site': site,
+                         'mutation_list': mutation_list}
+            self.sites.append(site_item)
+
+    def populate_tables(self, tables):
+        new_mutations = 0
+        for site_item in self.sites:
+            site = site_item['site']
+            site_id = tables.sites.add_row(**site._asdict())
+            for m in site_item['mutation_list']:
+                mut = m['mutation']
+                parent = mut.parent
+                if parent != tskit.NULL:
+                    parent += new_mutations
+                tables.mutations.add_row(site_id, mut.node, mut.derived_state,
+                                         parent=parent, metadata=mut.metadata)
+                if m['new']:
+                    new_mutations += 1
+
+    def place_mutations(self, tables, rng, discrete_sites=False):
+        # Insert a sentinel into the map for convenience.
+        map_position = np.hstack([self.rate_map.position, [tables.sequence_length]])
+        node_times = tables.nodes.time
+        for edge in tables.edges:
+            branch_start = node_times[edge.child]
+            branch_end = node_times[edge.parent]
+            branch_length = branch_end - branch_start
+            index = np.searchsorted(map_position, edge.left)
+            if map_position[index] > edge.left:
+                index -= 1
+            left = edge.left
+            right = 0
+            while right != edge.right:
+                right = min(edge.right, map_position[index + 1])
+                site_left = np.ceil(left) if discrete_sites else left
+                site_right = np.ceil(right) if discrete_sites else right
+                assert site_left <= site_right
+                assert map_position[index] <= left
+                assert right <= map_position[index + 1]
+                assert right <= edge.right
+                # Generate the mutations.
+                rate = self.rate_map.value[index]
+                mu = rate * (site_right - site_left) * branch_length
+                for _ in range(rng.poisson(mu)):
+                    position = rng.flat(site_left, site_right)
+                    if discrete_sites:
+                        position = np.floor(position)
+                    assert edge.left <= position
+                    assert position < edge.right
+                    time = rng.flat(branch_start, branch_end)
+                    site_pos = np.searchsorted([s['site'].position for s in self.sites],
+                                               position)
+                    if (site_pos == len(self.sites)
+                       or self.sites[site_pos]['site'].position != position):
+                        # new site
+                        ancestral_state = self.alleles[self.pick_allele()]
```

Comment:
I think it might be easier to separate the table maintenance from the internal data structures. This would simplify the object model as well. So, we if we stored a set of ``Site`` objects in the class (which holds a list of Mutations) and we did all of the work manipulating those, then once we're done we write these out to the Tables. It's hard and confusing work keeping two distinct views of the world consistent at the same time.

Reply:
I've found myself wanting to keep a few more things around, so have gone with your suggestion above of defining some new classes. Maybe it'll turn out we don't need those extra things, in which case using a Site would be perfect.